mom sharding to multi-server

bremanandjk opened the pull request:
**Describe Bug or Feature**
 To avoid the oversubscription of the default server, now pbs_mom would connect to any one of the random server (using the get_svr_shard_connection()).

**Describe Your Changes**
* Added new api named get_server_stream() to get the stream to connect the random server, this api would initialize the shard connection table to multiple-servers.
* If the initially connected server goes down, PBS_MOM would connect to the next available server for node info updates. 
* JOB updates would be sent to respective using the same get_server_stream() with the jobid as a hint to find its server.
* Added another API set_server_stream(), that helps to update the shard connection table to preserve the new incoming connection from the new stream descriptor to the server index appropriately.
* Updated the im_eof network down handler to handle the disconnection from the server and updates its state as DOWN. 

**Attach Test and Valgrind Logs/Output**
[Test_logs.txt](https://github.com/subhasisb/pbspro/files/4578320/Test_logs.txt)

<img width="1665" alt="MOM_Sharding Screen Shot " src="https://user-images.githubusercontent.com/17980660/81024197-b2616580-8e40-11ea-8258-52694d745751.png">
<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

bremanandjk commented:
@subhasisb and @suresh-thelkar, 
please share your reviews, Thanks

suresh-thelkar commented on src/lib/Libutil/misc_utils.c:
```diff
@@ -2196,12 +2196,23 @@ int
 	int i;
+	char *pc;
+	char *host1 = NULL;
+	char *host2 = NULL;
```
No need to initialise host1 and host2 to NULL.

bremanandjk replied:
ok, Done

suresh-thelkar commented on src/lib/Libutil/misc_utils.c:
```diff
@@ -2196,12 +2196,23 @@ int
+					*pc = '\0';
+				if ((pc = strchr(host2, (int)'.')) != NULL)
+					*pc = '\0';	
+				if (strcmp(host1, host2) == 0) 
```
Just a question. Are we only comparing short hostnames?

bremanandjk replied:
Yeah, since there could be possible that the caller may pass the other format.

suresh-thelkar commented on src/resmom/catch_child.c:
```diff
@@ -118,6 +118,7 @@ extern useconds_t	alps_release_jitter;
 #endif
 
 extern char		*path_hooks_workdir;
+extern int conn_slot;
```
Please align this variable with the formatting of the above variable declaration.

bremanandjk replied:
Done

suresh-thelkar commented on src/resmom/catch_child.c:
```diff
@@ -1937,6 +1938,111 @@ scan_for_exiting(void)
+				return -1;
+			}
+		}
+		set_new_shard_context(conn_slot);
```
Can we move the code between line numbers 1975 to 1987 to a separate initialisation function. This way these block of code is not called again when we call get_server_stream

bremanandjk replied:
Even if we create a new function, we would need to call that function as this place, and that function is not needed by any other parties. I am not sure about any gain in having separate functions. I think it's fine to have this code as it is, since "if condition" would guard against executing in the second occurrence.

suresh-thelkar replied:
As discussed if we create new function like initialise_shard_xyz(), the advantage is we can this function only once. We no need to check the if condition if (conn_slot == -1) each and every time as this is one time effort. Also please remove the following.
set_new_shard_context(conn_slot);
Please try to test with these modifications and if tests are passed you can go ahead with this.

bremanandjk replied:
I have added a separate function to initialize the virtual socket.
And here we cant remove the set_new_shard_context() function since this is needed to reset the connection. 

Please look into the modified changes.

suresh-thelkar commented on src/resmom/catch_child.c:
```diff
@@ -1937,6 +1938,111 @@ scan_for_exiting(void)
+		}
+		set_new_shard_context(conn_slot);
+		pfn_connect = internal_connect_mom;
+		stream = get_svr_shard_connection(conn_slot, -1 , jobid);	
```
Also can we rename conn_slot to virtual sock.

bremanandjk replied:
Done, renamed

suresh-thelkar commented on src/resmom/catch_child.c:
```diff
@@ -1937,6 +1938,111 @@ scan_for_exiting(void)
+		}
+		set_new_shard_context(conn_slot);
+		pfn_connect = internal_connect_mom;
+		stream = get_svr_shard_connection(conn_slot, -1 , jobid);	
```
Instead of -1 we can pass object type here right ?

bremanandjk replied:
Done, added JOB type

suresh-thelkar commented on src/resmom/catch_child.c:
```diff
@@ -1937,6 +1938,111 @@ scan_for_exiting(void)
+			return;
+		}
+		si.port = port;
+		shard_connection = (shard_conn_t **)get_conn_shards(conn_slot);
```
Can we move this line after get_svr_index(). Because if get_svr_index() itself fails then we no need to call get_conn_shards.

bremanandjk replied:
Done, moved the line.

suresh-thelkar commented on src/resmom/mom_comm.c:
```diff
@@ -2393,6 +2396,23 @@ im_eof(int stream, int ret)
 
+
+	for (j = 0; j < get_current_servers(); j++) {
+		shard_connection = (shard_conn_t **)get_conn_shards(conn_slot);
```
call get_conn_shards only if get_max_servers() > 1.

bremanandjk replied:
Done

suresh-thelkar commented on src/resmom/mom_comm.c:
```diff
@@ -2393,6 +2396,23 @@ im_eof(int stream, int ret)
+			continue;
+		}
+		if (stream == shard_connection[j]->sd) {
+			sprintf(log_buffer, "Server connection closed; changing connection state");
```
use snprintf.

bremanandjk replied:
Done

suresh-thelkar commented on src/resmom/mom_comm.c:
```diff
@@ -2393,6 +2396,23 @@ im_eof(int stream, int ret)
+				__func__, log_buffer);
+			shard_connection[j]->state = SHARD_CONN_STATE_DOWN;
+			shard_connection[j]->state_change_time = time(0);
+			set_new_shard_context(conn_slot);
```
Also do we need to reset the shard_context.

bremanandjk replied:
As we are resetting the connection state and we know that connected server got disconnected, it's better to reset to none.

suresh-thelkar commented on src/resmom/mom_server.c:
```diff
@@ -788,6 +794,10 @@ is_request(int stream, int version)
+	char	remote_server_name[NI_MAXHOST+1] = {'\0'};
+	struct	sockaddr_in	server_addr;
+	int errcode;
+	short	port;
```
Please align these variables with the above ones as early as possible.

bremanandjk replied:
Done.

bremanandjk commented:
@suresh-thelkar Thanks for your comments, I have addressed them.
please have a look.

subhasisb commented on src/resmom/mom_server.c:
```diff
@@ -805,7 +815,18 @@ is_request(int stream, int version)
+	memcpy((char *) &server_addr.sin_addr, &(addr->sin_addr), sizeof(server_addr.sin_addr));
+	server_addr.sin_port = addr->sin_port;
+	server_addr.sin_family = AF_INET;
+	if ((errcode = getnameinfo((struct sockaddr *) &server_addr, sizeof(struct sockaddr_in), remote_server_name,
```
name resolution look up for every request coming in would be very expensive - u should do it only if the server stream had already not been added

bremanandjk replied:
one ques:
Basically we are resolving the hostname to identify the **svr_index** of the incoming server.  If we are sure that all servers are running in the same host(the first version of multi-server), we could get the svr_index just by comparing the port of the incoming server alone. Is that fine?

bremanandjk replied:
I have moved the name resolution part to later that only when the stream is not available in shard connection table.

subhasisb commented on src/resmom/mom_server.c:
```diff
@@ -819,6 +840,8 @@ is_request(int stream, int version)
 	if (server_stream == -1)
 		send_hellosvr(stream);
 
+	set_server_stream(remote_server_name, port, stream);
```
error check?

bremanandjk replied:
Added error check.

subhasisb commented on src/resmom/mom_server.c:
```diff
@@ -1000,21 +1023,21 @@ is_request(int stream, int version)
 				}
 			}
-			if ((ret=is_compose(server_stream, IS_DISCARD_DONE)) != DIS_SUCCESS) {
+			if ((ret=is_compose(stream, IS_DISCARD_DONE)) != DIS_SUCCESS) {
```
space around =

bremanandjk replied:
Fixed this one and all other occurrences.

bremanandjk commented on src/resmom/catch_child.c:
```diff
@@ -1937,6 +1938,99 @@ scan_for_exiting(void)
+{
+	shard_conn_t **shard_connection = NULL;
+	int srv_index;
+	if (get_max_servers() > 1) {
```
Even I thought to avoid this switch, but here we are preserving the incoming connection in the shard_connection table, that table is not available for a single-server mode that is the reason I have guarded this code only to multi-server mode. 

For mom sharding, we are using this switch at 3 places, 
1) While choosing the server among multiple servers from the shard connection table. 
2) Preserving the incoming server stream in the shard connection table. 
3) If the connection from the server drops, need to change the connection state in the table. 

These 3 situations are mostly related to using a shard connection table that is not applicable for a single mode. so I couldn't find a way of combining these codes in one code path.

subhasisb replied:
ok so we can make the shard table for single server mode also, so that there is only one connection in the shard table.

bremanandjk replied:
I have tried this change for commands integration, sharing my observations here
 
1) Job submissions to the server are working. (By having shard_connection table with 1 entry in it)
2) But the scheduler won't work now until Sched sharding PR merged in. (All the sched -> server IFL connection should pass the virtual socket after initializing shard table)
3) Even with the addition of MoM sharding cant run automation tests since the scheduler can't schedule proper nodes. 
4) Also, differences in the errno and errmsg cause the PTL test to fail. (Since we introduced new errno for multi-server mode.)

My changes are here: https://github.com/bremanandjk/pbspro/commit/f886d03c0ae4cfcca27b392ef729e57b43ed3710

So, it would be better to merge Mom sharding and Scheduler sharding with this switch. Once all changes are in, we should easily remove this switch and make all the necessary test case changes for the same code path.

bremanandjk commented on src/lib/Libutil/misc_utils.c:
```diff
@@ -2196,12 +2196,23 @@ int
+				if ((!instance->name) || (!pbs_conf.psi[i]->name))
 					return  -1;
-				if (strcmp(instance->name, pbs_conf.psi[i]->name) == 0) 
+				host1 = strdup(instance->name);
```
I have changed code to compare two hostnames using strchr and compare it to that length.

bremanandjk commented:
I am seeing some extra commits in this PR after rebasing with the mainline master, I have rebased and squashed my commit in new PR. https://github.com/subhasisb/pbspro/pull/86

bremanandjk commented:
Raised another PR, closing this one.
